### PR TITLE
refactor(package): remove `node-sass` (`peerDependencies`)

### LINKED ISSUE #532: Consider removing peerDependency node-sass and enforcing it at runtime instead
Hey 👋 

Create React App users have been [very](https://github.com/facebookincubator/create-react-app/issues/78) [vocal](https://github.com/facebookincubator/create-react-app/issues/2498) about wanting to have Sass supported out of the box.

My [personal concerns](https://github.com/facebookincubator/create-react-app/issues/78#issuecomment-245727905) about `node-sass` haven't changed: it’s unfortunate to impose downloading a binary on each our user, and especially unfair to subject them to build issues like https://github.com/sass/node-sass/issues/2146 if they’re not even planning to rely on Sass.

But since the demand for Sass isn’t going away, I’d like to make it opt-in for our users. The ideal user experience would be that they don’t need to download `node-sass` by default and don’t see any peer dependency warnings, but if they attempt to import a Sass file, they will see an error message telling them to install a compatible version of `node-sass`.

The biggest roadblock to this experience is that `sass-loader` will just crash without `node-sass` installed. We could prevent this by putting our own custom loader in front of it, but this still doesn’t solve the scary `peerDependency` warning coming from our package if it depends on `sass-loader` but not `node-sass`. And, as I mentioned above, we don’t want to force our users to install `node-sass` unless they actually plan to use it.

Here is how `sass-loader` works today:

* It has a `peerDependency` on `node-sass`
* If `node-sass` doesn't exist, `sass-loader` crashes

Here is my proposal for how it could work:

* It has no `peerDependency` on `node-sass`
* The import of `node-sass` is wrapped in a `try` / `catch`
* If `node-sass` is missing or `require('node-sass').info` doesn’t match the compatible range, the loader issues a compilation error telling the user to install a compatible version of `node-sass`

This allows tools like Create React App to depend on `sass-loader` by default without creating hurdles for users who don’t want it. What do you think? We’d be happy to send a PR ourselves if you can get behind this change in behavior.

### PR DESCRIPTION
**Pull request by Timer**
Fixes #532

Let me know what more is required. Doc changes, fixing coverage, etc.

image

image

image

### CODE REVIEW

File: lib/loader.js [hunk 1]
```diff
@@ -12,7 +11,21 @@ const pify = require("pify");
...
 const threadPoolSize = process.env.UV_THREADPOOL_SIZE || 4;
-const asyncSassJobQueue = async.queue(sass.render, threadPoolSize - 1);
+let asyncSassJobQueue = null;
+let loadingError = null;
+
+try {
+    const sass = require("node-sass");
+    const sassVersion = require("node-sass/package.json").version;
+
+    if (/^4[.]/.test(sassVersion)) {
```
Comment (evilebottnawi) [inline lib/loader.js]: I think better to check `>= 4`, maybe in future we can have `5`

Comment (michael-ciniawsky) [reply lib/loader.js]: Yep, please check for greater than  

```js
if (Number(sassVersion[0]) >= 4) {}
```

File: lib/loader.js [hunk 2]
```diff
@@ -12,7 +11,21 @@ const pify = require("pify");
...
+        asyncSassJobQueue = async.queue(sass.render, threadPoolSize - 1);
+    } else {
+        loadingError = new Error("The installed version of \`node-sass\` is not compatible (expected: 4.x, actual: " + sassVersion + ").");
```
Comment (michael-ciniawsky) [inline lib/loader.js]: `throw` here and catch the `{Error}` in (L27+) instead of an assignment please

File: lib/loader.js (hunk 1 above)
Comment (Timer) [reply lib/loader.js]: What if we're not compatible with the `node-sass` v5 API?

File: lib/loader.js (hunk 2 above)
Comment (Timer) [reply lib/loader.js]: Good point. This will work better when we add support for things other than `node-sass` or something else bombs out.

**Comment by Timer** [comment]
Updated with concerns.

File: lib/loader.js [hunk 3]
```diff
@@ -12,7 +11,21 @@ const pify = require("pify");
 // fs tasks when running the custom importer code.
 // This can be removed as soon as node-sass implements a fix for this.
 const threadPoolSize = process.env.UV_THREADPOOL_SIZE || 4;
```
Comment (evilebottnawi) [inline lib/loader.js]: Let move this variables before `asyncSassJobQueue = async.queue(sass.render, threadPoolSize - 1);`, it don't have sense in top of file

Comment (Timer) [reply lib/loader.js]: Makes sense. I should've thought of this. 😝

File: lib/loader.js [hunk 4]
```diff
@@ -35,6 +49,15 @@ function sassLoader(content) {
         throw new Error("Synchronous compilation is not supported anymore. See https://github.com/webpack-contrib/sass-loader/issues/333");
     }
 
+    if (asyncSassJobQueue == null) {
```
Comment (evilebottnawi) [inline lib/loader.js]: Last note :smile: Seems we can avoid this block, just throw error in `catch` instead `loadingError = e;`

Comment (Timer) [reply lib/loader.js]: Is it safe to throw an error at require-time? I figured you'd want them surfaced as webpack compiler errors -- I'm not sure how this works.

Comment (evilebottnawi) [reply lib/loader.js]: @Timer Yep, it is not compilation error. It does not make sense because all attempts to compile `sass/scss` return error in callback. I think better throw error as fast as possible. @michael-ciniawsky what do you think about this?

Comment (Timer) [reply lib/loader.js]: Wouldn't webpack's `bail` option cause it to stop on the first? I'm no expert here, just asking questions.

Comment (michael-ciniawsky) [reply lib/loader.js]: I also think this block is redundant with the `try/catch` (L13+) and version check, but not 💯 😛  

```diff
- let loadingError = null;

try {
    const sass = require("node-sass");
    const sassVersion = require("node-sass/package.json").version;

    if (Number(sassVersion[0]) >= 4) {
        // This queue makes sure node-sass leaves one thread available for executing
        // fs tasks when running the custom importer code.
        // This can be removed as soon as node-sass implements a fix for this.
        const threadPoolSize = process.env.UV_THREADPOOL_SIZE || 4;

        asyncSassJobQueue = async.queue(sass.render, threadPoolSize - 1);
+       // Maybe not needed (see comment)       
+       if (asyncSassJobQueue === null) {
+           throw new Error("Please install `node-sass >= 4.x` to use `sass-loader`.")
+       }
    } else {
        throw new Error("The installed version of \`node-sass\` is not compatible (expected: >=4.x, actual: " + sassVersion + ").");
    }
} catch (err) {
+  // Catches 'both' (first thrown) {Error} thrown in the try block 
+   callback(err);

+   return;
}

...

-    if (asyncSassJobQueue == null) {
-        if (loadingError) {
-            callback(loadingError);
-            return;
-        }
-        callback(new Error("Please install `node-sass >=4.x` to use `sass-loader`."));
-        return;
-    }
```

But what's the exact reason for `asyncQueue === null`, in case `node-sass < 4` ? And isn't this basically a redundant (the same check) as `if (Number(sassVersion[0])) >= 4)` again ? Please elaborate what could possibly fail when queueing `sass.render` for thread pooling, since I'm not that familiar with `node-sass` :)

Comment (Timer) [reply lib/loader.js]: Callback isn't available where you've added it, that's where the package itself is being evaluated.

AFAIK, it's generally poor practice to make a require throw an exception, but this might be normal practice in the webpack world.

The `loadingError` is to store the error until the plugin is used at compile time.

Comment (michael-ciniawsky) [reply lib/loader.js]: Ooops... 😛you're right. What alternatives do we have to make this less confusing and more compact ?

Comment (Timer) [reply lib/loader.js]: I'll come up with something.

**Comment by Timer** [comment]
@michael-ciniawsky I went ahead and moved the loading of a compiler into its own file and created placeholder logic for expanding beyond one compiler. This keeps the main `loader.js` clean while allowing it to be extensible going forward.

Let me know what you think!

**Comment by Timer** [comment]
> Can we add tests?

Sure! I just want to make sure the code is accepted and no functionality majorly changes before I write the test cases. 😄

**Comment by evilebottnawi** [comment]
@michael-ciniawsky friendly ping

**Comment by Timer** [comment]
@michael-ciniawsky could you please see if you agree with the changes? 😄

File: lib/loader.js [hunk 5]
```diff
@@ -35,6 +30,13 @@ function sassLoader(content) {
         throw new Error("Synchronous compilation is not supported anymore. See https://github.com/webpack-contrib/sass-loader/issues/333");
     }
 
+    const compilerError = asyncJobQueue.error;
+
+    if (compilerError) {
```
Comment (jhnns) [inline lib/loader.js]: No need to do that check. If the error was just thrown, webpack would report the loader error accordingly.

File: lib/compilers/index.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+            return { error: false, reason: null, value };
+        } catch (e) {
+            if (firstError == null) {
+                firstError = e;
+            }
+        }
+    }
+    // Return the error encountered while loading a compiler, otherwise say
+    // there was no compiler found.
+    return {
```
[the remark this answers is not in the document]
Comment (Timer) [reply lib/compilers/index.js]: AFAIK throwing on `require` is being a bad neighbor, so we store the error until the loader is invoked.

File: lib/compilers/index.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+function getCompilerQueue() {
+    // Stores the first error encountered while loading a compiler (e.g. invalid version)
+    // This error will be ignored if one of the compilers succeed (since we will eventually support multiple).
+    let firstError;
+
+    for (const compiler of compilers) {
+        try {
+            const value = compiler.getCompilerQueue();
+
+            if (value == null) {
```
[the remark this answers is not in the document]
Comment (Timer) [reply lib/compilers/index.js]: Fair enough.

**Comment by Timer** [comment]
I believe `dart-sass` support would be better in a PR after this. I toyed with the idea (see tree at d2dae27210a01b7be9e23ee0386743335d0e75ea) but I think it'll require more significant changes (such as the error message massaging and fixing library-specific quirks).

Let's keep this PR focused to removing the peer dep.

**Comment by Timer** [comment]
@jhnns I've greatly simplified this PR; can I ask for another review?

**Comment by Timer** [comment]
Remaining tests should pass after https://github.com/webpack-contrib/sass-loader/pull/514 is merged.

**Comment by Timer** [comment]
Hi! Anything we can do to help this along?

We'd really like to add Sass support in Create React App, but we need this released first.

File: .nycrc [hunk 8]
```diff
@@ -7,8 +7,8 @@
         "lib/**/*.js"
     ],
     "lines": 97,
-    "statements": 91,
+    "statements": 97,
     "functions": 100,
-    "branches": 89,
+    "branches": 91,
```
Comment (evilebottnawi) [inline .nycrc]: Can we return old values and add more tests?

Comment (evilebottnawi) [reply .nycrc]: /cc @Timer

Comment (Timer) [reply .nycrc]: Just to clarify, you want me to reduce target test coverage and add then add more tests?

I increased the value(s) here with the tests I have already added -- I did not decrease them.

**Comment by jhnns** [comment]
LGTM 👍 thank you 👏 

Travis is failing right now since it's trying to run webpack@4 with node 4 which doesn't work... I'll try to get #514 merged so that our CI is working again.

**Comment by jhnns** [comment]
I will release this tomorrow together with #479 #555

**Comment by duivvv** [comment]
this is a massive PR, lots of people will be happy :)

File: package.json [hunk 9]
```diff
@@ -54,7 +55,6 @@
     "node": ">= 6.9.0 || >= 8.9.0"
   },
   "peerDependencies": {
-    "node-sass": "^4.0.0",
```
Comment (track0x1) [inline package.json]: Why did you remove `node-sass` as a peerDependency when `sass-loader` explicitly requires a version?

Comment (nex3) [reply package.json]: `sass-loader` requires *either* `node-sass` or `sass`.  There's no way to express that as a peer dependency without having npm complain about one or the other missing.

Comment (track0x1) [reply package.json]: Ah! Thank you. Didn't know you could use either.
